compiler/rustc_lint/src/redundant_sizedness_bound.rs

```diff
@@ -0,0 +1,240 @@
+use rustc_errors::Applicability;
+use rustc_hir::def_id::{DefId, DefIdMap};
+use rustc_hir::{
+    BoundPolarity, GenericBound, Generics, PolyTraitRef, TraitBoundModifiers, WherePredicateKind,
+};
+use rustc_middle::ty::{ClauseKind, PredicatePolarity};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::symbol::Ident;
+
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `redundant_sizedness_bound` lint detects redundant sizedness bounds applied to type parameters that are already
+    /// otherwise implied.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #![feature(sized_hierarchy)]
+    /// use std::marker::MetaSized;
+    /// // `T` must be `Sized` due to the bound `Clone`, thus `?Sized` is redundant.
+    /// fn f<T: Clone + ?Sized>(t: &T) {}
+    /// // `T` is `Sized` due to `Clone` bound, thereby implying `MetaSized` and making the explicit `MetaSized` bound redundant.
+    /// fn g<T: MetaSized + Clone>(t: &T) {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Sizedness bounds that have no effect as another bound implies a greater degree of sizedness are potentially misleading
+    /// This lint notifies the user of such redundant bounds.
```

Comment:
Similarly, I'd probably remove the wording here about "degree of sizedness" since, until we stabilize `MetaSized` and friends, we don't have any degrees of this in Rust.  We can say instead, e.g., that another bound implies `Sized` or words to similar effect.